### ISSUE TITLE
Annual and hourly results differ for P2H in hydrogen, natural gas and heat network balances

### ISSUE THREAD

**Issue by kndehaan** [quintel/etengine#1486]
The yearly and annual output of the following technologies don't add up:
- P2H heat pump HT
- P2H heat pump MT
- P2H boiler HT
- P2H boiler MT

This was noted while creating mekkos for the heat networks: https://github.com/quintel/etsource/pull/3164#issuecomment-2553358558

In [this scenario](https://energytransitionmodel.com/saved_scenarios/19071) there's installed capacity and flows of all above technologies, as well as flows from the backup burners HT and MT (there was suspicion that the annual and hourly flows of these technologies didn't match as well, however this is not the case). 

Querying the annual and output of steam hot water gives a consistent `0.00178` difference between annual and hourly flows. For instance querying:

```
EACH(
  V(energy_heat_flexibility_p2h_heatpump_ht_electricity, output_of_steam_hot_water) / BILLIONS,
  DIVIDE(
    SUM(V(energy_heat_flexibility_p2h_heatpump_ht_electricity, steam_hot_water_output_curve)) * 3.6,
    MILLIONS
  )
)
```
gives:

```
[
  5.150360382499293,
  5.148578382499292
]
```

Would be useful to dive deeper into this issue at some point. Notifying @mabijkerk and @kaskranenburgQ

**Comment by mabijkerk** [comment]
As indicated by @kaskranenburgQ in https://github.com/quintel/etsource/issues/3080#issue-2380643555, for the IP2024 National Drivers [scenario](https://beta.energytransitionmodel.com/saved_scenarios/5126), both the P2H heat pump and boiler on the heat network show a difference in hourly and annual values:

```
EACH(
  V(energy_heat_flexibility_p2h_boiler_ht_electricity,output_of_steam_hot_water),
  SUM(V(energy_heat_flexibility_p2h_boiler_ht_electricity,steam_hot_water_output_curve))*MJ_PER_MWH,
  V(energy_heat_flexibility_p2h_heatpump_ht_electricity,output_of_steam_hot_water),
  SUM(V(energy_heat_flexibility_p2h_heatpump_ht_electricity,steam_hot_water_output_curve))*MJ_PER_MWH
)

[
  33,278,159,838.49849,
  33,271,031,838.49849,
  16,639,079,919.249239,
  16,635,515,919.249243,
]
```

**Comment by kndehaan** [comment]
It would be good for @noracato or @louispt1 to have a look at the functionality of the [power_to_heat_adapter](https://github.com/quintel/etengine/blob/master/app/models/qernel/merit_facade/power_to_heat_adapter.rb) as a next step in solving this issue. We can focus on this [simple scenario](https://my.energytransitionmodel.com/saved_scenarios/19850) created from a blank nl2019, where 1000 MW capacity of the P2H boiler is set to feed into the MT network. Then we have the following disbalance between annual and hourly steam hot water output:

```
EACH(
  V(energy_heat_flexibility_p2h_boiler_mt_electricity, output_of_steam_hot_water),
  SUM(V(energy_heat_flexibility_p2h_boiler_mt_electricity, steam_hot_water_output_curve))*MJ_PER_MWH
  )
```

```
[
  31,220,640,000.000004,
  31,217,076,000.000008,
]
```
The difference is 3,564,000 MJ. Note that when running the scenario locally and steam hot water output conversion to 1, availability to 1, and typical input capacity equal to heat output capacity, the difference between the annual and hourly calculations is a neat 3,600,000 MJ (could this be a clue?). 

From manual calculations based on installed heat output capacity as set in the input, and the future year full load hours (how are full load hours determined here exactly?), the annual heat production should be 31,220,640,000 MJ. Would this argue that the result of the annual calculations are correct, and that the hourly calculation is wrong?

I will discuss this further with either @noracato or @louispt1 to have a look together in solving this issue.

**Comment by kndehaan** [comment]
As discussed with @noracato a next step was to verify whether this issue occurs for both the P2H technologies with `merit_order.subtype = power_to_heat_industry` (all P2H i.c.w. gas or hydrogen boiler in agriculture and industry) and technologies with `merit_order.subtype = generic` (all P2H for district heating). This is needed since the `power_to_heat_industry` types rely on the `PowerToHeatAdapter`, whereas the `generic` rely on the general `FlexAdapter`, if I say this correctly. 

It seems that the problem occurs for both P2H merit order subtypes. See [this scenario](https://my.energytransitionmodel.com/saved_scenarios/19896) where a 1000 MW heat output capacity for the agriculture P2H and 'regular' P2H boiler for district heating are set.

```
EACH(
  V(agriculture_flexibility_p2h_network_gas_electricity, output_of_useable_heat),
  SUM(V(agriculture_flexibility_p2h_network_gas_electricity, useable_heat_output_curve)) * MJ_PER_MWH,
  V(energy_heat_flexibility_p2h_boiler_ht_electricity, output_of_steam_hot_water),
  SUM(V(energy_heat_flexibility_p2h_boiler_ht_electricity, steam_hot_water_output_curve)) * MJ_PER_MWH
)

[
  30,556,171,826.78359,
  30,568,233,095.34177,
  31,220,640,000.000004,
  31,217,076,000.000008
]

This means that it seems not to only be a problem in the `PowerToHeatAdapter`... @noracato I hope this information is of use for you. Please let me know if I can be of any help in further solving this issue.

**Comment by mabijkerk** [comment]
I had AI investigate this further and I came with the following.

## Root cause analysis: annual vs hourly mismatch for P2H

There are two distinct mechanisms, and the numbers from the query in the issue confirm the first one *exactly*.

### 1. District heating P2H: the last hour of electricity never becomes heat

These nodes live in **two** merit orders: flex consumer (`temperature_based_p2h`) in the electricity MO, and `must_run` producer in the heat network MO with `heat_network_ht.group = self: electricity_input_curve`. The causality loop calculates the heat network **before** electricity each hour ([causality.rb#L96-L102](https://github.com/quintel/etengine/blob/master/app/models/qernel/plugins/causality.rb#L96-L102)), so the heat side can't know this hour's electricity load yet. To handle that, [self_curves.rb#L183-L192](https://github.com/quintel/etengine/blob/master/app/models/qernel/merit_facade/self_curves.rb#L183-L192) reads the electricity curve with a **1-hour offset**: heat hour h = electricity hour h−1, and heat hour 0 = 0.0.

Consequence: electricity consumed in hour 8759 never appears in `steam_hot_water_output_curve`. But the node's annual demand ([flex_adapter.rb#L116-L133](https://github.com/quintel/etengine/blob/master/app/models/qernel/merit_facade/flex_adapter.rb#L116-L133)) is the *full* electricity consumption including hour 8759, and annual `output_of_steam_hot_water` = demand × slot conversion. So **annual exceeds the hourly sum by exactly the last hour's load**.

The numbers from the IP2024 National Drivers scenario verify this precisely:

- Boiler: diff = 7,128,000 MJ = 1980 MWh heat = its full installed capacity for one hour (1989.9 MWh electricity × 0.995).
- Heat pump: diff = 3,564,000 MJ = 990 MWh = full capacity × 1 hour.
- Both work out to 4668 full-load hours — the number of hours below the 17 °C `temperature_cutoff`. Both ran flat-out every available hour, including the cold final hour of the year, whose heat is dropped.

The kicker: **this exact problem was already fixed once** — for CHPs. [`inject_self_shares!` in producer_adapter.rb#L42-L71](https://github.com/quintel/etengine/blob/master/app/models/qernel/merit_facade/producer_adapter.rb#L42-L71) (from quintel/etengine#1175) shaves the heat output conversion by the lost last hour. But it guards on `group == 'self:electricity_output_curve'` and electricity subtype `:dispatchable`, so the P2H case (`electricity_input_curve`, flex) silently falls through.

### 2. Industry/agriculture P2H (the hybrid heat burner gas/hydrogen mismatch): energy stranded in the buffer

`power_to_heat_industry` nodes (e.g. `industry_chemicals_other_flexibility_p2h_hydrogen_electricity`) use the storage-based [power_to_heat_adapter.rb](https://github.com/quintel/etengine/blob/master/app/models/qernel/merit_facade/power_to_heat_adapter.rb). There:

- Annual demand = `participant.production` = all electricity **charged into** the reserve over the year (Merit `Flex::Base#production` sums negative loads).
- The hourly `heat_output_curve` is filled only by what **decays out** of the reserve. Whatever is still stored at hour 8759 (these nodes have `storage.volume = 50 MWh/unit`) is counted annually but never emitted hourly. Charging in hour h can also only decay out from h+1 onwards, so hour 0 contributes nothing.
- Bonus inconsistency in [`reserve_decay` (power_to_heat_adapter.rb#L74-L88)](https://github.com/quintel/etengine/blob/master/app/models/qernel/merit_facade/power_to_heat_adapter.rb#L74-L88): the lambda computes `wanted = decay.get(point) / conversion` but returns the *unconverted* `decay.get(point)` to the reserve — so conversion losses (the 0.005 loss slot) are never actually withdrawn from the buffer, while the annual graph does apply the 0.995 conversion.

This is what hits the hybrid heat burners: [unmet_demand_curves.rb#L85-L92](https://github.com/quintel/etengine/blob/master/app/models/qernel/merit_facade/unmet_demand_curves.rb#L85-L92) makes the hydrogen/gas backup burner cover `subtraction_profile − heat_output_curve` hourly. Since the hourly P2H curve sums to less than its annual output, the hourly gas/hydrogen consumption disagrees with the annual graph flows — the imbalance observed here.

### Fix direction

1. **District heating**: extend `inject_self_shares!` to cover must_run heat producers backed by a flex electricity participant with `self: electricity_input_curve` (using the input-filtered value at frame 8760, analogous to the CHP case). The boiler works as-is (`output.loss = elastic` absorbs the shaved conversion), but `energy_heat_flexibility_p2h_heatpump_ht_electricity.ad` has **no loss slot** (`output.steam_hot_water = 1.0`) — it needs `output.loss = elastic` added in ETSource, otherwise reducing the conversion breaks the node balance. Same check needed for the MT variants.
2. **Industry P2H**: in `PowerToHeatAdapter#inject!`, treat the year-end reserve leftover as loss (shave the `useable_heat` conversion so annual output matches `SUM(heat_output_curve)` — these nodes all have elastic loss slots), and decide the intended semantics of the decay lambda (return `wanted` so the reserve pays for conversion losses).

Note these mismatches are conceptually different: in case 1 the energy is genuinely consumed but never delivered (a real model artifact best booked as loss); in case 2 part of it is just sitting in the buffer on Dec 31. The correction is the same trick either way — make the annual conversions consistent with the hourly curve.

**Comment by mabijkerk** [comment]
On the ETEngine branch [mismatch-fix](https://github.com/quintel/etengine/tree/mismatch-fix) I added the change AI suggested for issue 2 - this is just meant for inspiration! Since I was already in the session I felt a bit of a waste not to have it write the changes as well.

I also had it add the change for issue 1. I'm a bit skeptical about it changing the efficiency of steam hot water to slightly below 100% (for the heat pumps that have steam hot water output conversion of 1.0) to have the elastic loss absorp the difference - but apparently we already do this for CHPs as well (though none of them will have 100% efficiency). This required a minor change on the [mismatch-fix](https://github.com/quintel/etsource/tree/mismatch-fix) branch on ETSource as well. Small discussion might be needed to see if this indeed the best solution (though perhaps having an insignificant loss is not too bad of a solution).

In any case, this root cause analysis and potential solutions require further follow-up. **I did not yet check or test this thoroughly in any way!**

Ideally, we would add a spec that verifies consistency between annual and hourly results.
